pkg/reconciler/v1alpha1/taskrun/taskrun_test.go

```diff
@@ -1542,17 +1547,168 @@ func TestUpdateStatusFromBuildStatus(t *testing.T) {
 			StartTime:      &startTime,
 			CompletionTime: &completionTime,
 		},
+		tr: &v1alpha1.TaskRun{},
 	}} {
 		t.Run(tc.name, func(t *testing.T) {
-			taskRun := &v1alpha1.TaskRun{}
-			updateStatusFromBuildStatus(taskRun, tc.buildStatus)
-			if d := cmp.Diff(taskRun.Status, tc.expectedStatus, ignoreLastTransitionTime); d != "" {
+			//taskRun := &v1alpha1.TaskRun{}
+			updateStatusFromBuildStatus(tc.tr, tc.buildStatus)
+			if d := cmp.Diff(tc.tr.Status, tc.expectedStatus, ignoreLastTransitionTime); d != "" {
 				t.Errorf("-want, +got: %v", d)
 			}
 		})
 	}
 }
 
+func TestRetryFailedTaskRun(t *testing.T) {
+
+	completed := corev1.ContainerState{
+		Terminated: &corev1.ContainerStateTerminated{ExitCode: 0, Reason: "success"},
+	}
+
+	failed := corev1.ContainerState{
+		Terminated: &corev1.ContainerStateTerminated{ExitCode: 127, Reason: "oh-my-lord"},
+	}
+	startTime := metav1.NewTime(time.Date(2018, time.November, 10, 23, 0, 0, 0, time.UTC))
+	completionTime := metav1.NewTime(time.Date(2018, time.November, 10, 23, 8, 0, 0, time.UTC))
+
+	taskRun := tb.TaskRun("test-taskrun-failed-with-retry", "foo",
+		tb.TaskRunSpec(
+			tb.TaskRunTaskRef(simpleTask.Name),
+			tb.TaskRunRetries(1),
+		),
+		tb.TaskRunStatus(tb.Condition(duckv1alpha1.Condition{
+			Type:   duckv1alpha1.ConditionSucceeded,
+			Status: corev1.ConditionUnknown}),
+		))
+
+	buildStatus := buildv1alpha1.BuildStatus{
+		StartTime:      &startTime,
+		CompletionTime: &completionTime,
+		StepStates: []corev1.ContainerState{
+			completed,
+			failed,
+		},
+		Conditions: []duckv1alpha1.Condition{{
+			Type:    duckv1alpha1.ConditionSucceeded,
+			Status:  corev1.ConditionFalse,
+			Reason:  "Build failed",
+			Message: "Build failed",
+		}},
+		Cluster: &buildv1alpha1.ClusterSpec{
+			Namespace: "default",
+			PodName:   "im-am-the-pod",
+		},
+	}
+
+	retries := v1alpha1.TaskRunStatus{
+		Conditions: []duckv1alpha1.Condition{{
+			Type:    duckv1alpha1.ConditionSucceeded,
+			Status:  corev1.ConditionFalse,
+			Reason:  "Build failed",
+			Message: "Build failed",
+		}},
+		PodName:        "im-am-the-pod",
+		StartTime:      &startTime,
+		CompletionTime: &completionTime,
+		Steps: []v1alpha1.StepState{
```

Comment:
nit: to align with rest of tests it might be better to have `{{` to indicate array

```
Steps: []v1alpha1.StepState{{
	ContainerState: *completed.DeepCopy(),
	},{
	ContainerState: *failed.DeepCopy(),
}},
```